### DISCUSSION TITLE
Validating computed fields: please add to documentation

### DISCUSSION THREAD

**Discussion by vineetrajasekhar** [pydantic/pydantic#10098] [Pydantic V2]
Hi, 

(**TL;DR:** Users may (incorrectly but understandably) assume that `computed_field`s are evaluated when a model's instance is created, and expect validation logic in a `computed_field` definition to prevent invalid object creation. Please consider updating the docs to clearly warn that `computed_field` must be separately validated during object creation, e.g. by `model_validator`.)

I'm loving Pydantic V2 - thank you for all the hard work that's evidently gone into it. I particularly find `computed_field` very valuable. 

However, an assumption **I (incorretly) made** about its behaviour got me into trouble when transitioning a project from V1 to V2 recently. I assumed that `computed_field` runs when a model is instantiated, thereby ensuring that any validation logic and exceptions written into the `computed_field` definition would run when a model's instance is created, rather than later when that instance is used. 

For example: 

```python
from pydantic import BaseModel, Field, computed_field, model_validator
from typing_extensions import Annotated, Self

class SkinnyRectangle(BaseModel):
    """
    A rectangle whose length is at least 1.5 times its width.
    """

    width: Annotated[float, Field(gt=0, strict=True)]
    length: Annotated[float, Field(gt=0, strict=True)]

    @computed_field
    @property
    def aspect_ratio(self) -> float:
        """
        (My original expectation)
        
        Computes the rectangle's aspect ratio AND raises an exception if the 
        computed value is invalid.
        """
        ratio = self.length / self.width
        
        if ratio <= 1.5:
            raise ValueError("Aspect ratio  must be at least 1.5")
        return ratio

def main():
    valid_width, valid_length = (1, 1.5)
    invalid_width, invalid_length = (1, 1.2)

    # I don't expect this to raise any exceptions, and it doesn't. 
    valid_rectangle = SkinnyRectangle(width=valid_width, length=valid_length)

    # I EXPECT this to raise an exception, but, worryingly, it DOESN'T
    invalid_rectangle = SkinnyRectangle(width=invalid_width, length=invalid_length)

```

Luckily my test suite alerted me to the fact that invalid models were getting through and I worked out that, in addition to the `computed_field`, I also needed a `model_validator` to subsequently check that my computed field exhibited valid characteristics, like so:

```python
from pydantic import BaseModel, Field, computed_field, model_validator
from typing_extensions import Annotated, Self

from logs import log


class SkinnyRectangle(BaseModel):
    """
    A rectangle whose length is at least 1.5 times its width.
    """

    width: Annotated[float, Field(gt=0, strict=True)]
    length: Annotated[float, Field(gt=0, strict=True)]

    @computed_field
    @property
    def aspect_ratio(self) -> float:
        """
        (My current understanding:)

        Computes the rectangle's aspect ratio AFTER object creation, only when this 
        attribute is accessed (during object use)
        """
        ratio = self.length / self.width
        return ratio

# Model validator added to examine the above computed field 

    @model_validator(mode="after")
    def check_aspect_ratio(self) -> Self:
        ratio = self.aspect_ratio
        if ratio <= 1.5:
            raise ValueError("Aspect ratio must be at least 1.5")
        return self


def main():
    valid_width, valid_length = (1, 1.5)
    invalid_width, invalid_length = (1, 1.2)

    # I don't expect this to raise any exceptions, and it doesn't.
    valid_rectangle = SkinnyRectangle(width=valid_width, length=valid_length)

    # Now that I have a model_validator implemented to check aspect_ratio, 
    # the following instance creation fails and the expected exception is raised
    invalid_rectangle = SkinnyRectangle(width=invalid_width, length=invalid_length)


if __name__ == "__main__":
    main()

```

I now understand (from #8865) that this behaviour is a design choice, and nothing in the [documentation for `computed_field`](https://docs.pydantic.dev/2.8/api/fields/#pydantic.fields.computed_field ) necessarily implies the behaviour I assumed. 

I do think (from #9346, for example) that I am probably not alone in expecting validation behaviour from `computed_field`, given most people (like me) transitioning from V1 to V2 are likely converting `root_validator`s to `computed_field`s and may fail to expect `computed_field` to need a standalone validator. 

**Can I ask that the docs be updated to clearly warn that `computed_field` does not run on model creation, and that a separate validator function is needed to validate `computed_field`s?**

Thank you very much!
Vineet

**P.S.** I do think it might be nice to be able to tersely imbue `computed_field` with validation  behaviour, perhaps with a `compute_on_creation=True` kwarg passed to the `compute_field` decorator. However I haven't thought through the implications, and so have not raised a feature request yet.

**Comment by vineetrajasekhar** [comment]
Approximately related PR: #9131

**Comment by vineetrajasekhar** [comment]
Related PR: #6545

**Comment by Viicos** [comment]
I think the [documentation](https://docs.pydantic.dev/latest/concepts/fields/#the-computed_field-decorator) is pretty clear on what `computed_field` does:

> The `computed_field` decorator can be used to include `property` or `cached_property` attributes when serializing a model or dataclass.

I believe the right path is to emphasis that `computed_fied` **only** affects serialization and JSON Schema, and appart from this it behaves as a regular property (the thing is we don't want to enumerate every potential misunderstanding from users).

**Reply by vineetrajasekhar** [reply]
Hi @Viicos,

I agree that amending the docs to delimit the scope/applicability of `computed_field` would be worthwhile, and would appreciate seeing this change made.

I also agree that enumerating every user misunderstanding would risk turning the documentation into a mess of rules and exceptions thereof. However, in view of Pydantic's primary (or perhaps most well-known) use-case as a validation engine, perhaps unambiguously declaring validation as a "non-goal" of `computed_field` might be worth considering.

Thank you,
Vineet

**Reply by Viicos** [reply]
I opened https://github.com/pydantic/pydantic/pull/10103 this afternoon. Happy to get review if you want to.
